src/analyses/pthreadAnalysis.ml

```diff
@@ -0,0 +1,1359 @@
+(** Tracking of pthread lib code. Output to promela. *)
+
+open Prelude.Ana
+open Analyses
+open Cil
+open Deriving.Cil
+open BatteriesExceptionless
+open Option.Infix
+
+class uniqueVarPrinterClass =
+  object (self)
+    inherit defaultCilPrinterClass as super
+
+    method! pVar (v : varinfo) =
+      text v.vname ++ chr '_' ++ text (string_of_int v.vid)
+
+    method! pExp () =
+      function
+      | Const (CInt64 (i, _, _)) ->
+          (* Fix the constants with long/unsigned suffixes, e.g. 1LL *)
+          text @@ string_of_int @@ Int64.to_int i
+      | x ->
+          super#pExp () x
+  end
+
+let printer = new uniqueVarPrinterClass
+
+module Flags = struct
+  type t =
+    | AssumeSuccess
+    | IgnoreAssigns
+
+  let all = [ AssumeSuccess; IgnoreAssigns ]
+
+  let default_value = function AssumeSuccess -> true | IgnoreAssigns -> true
+
+  let description = function
+    | AssumeSuccess ->
+        "Assume that all POSIX pthread functions succeed."
+    | IgnoreAssigns ->
+        "Ignors any assigns in POSIX programs"
+
+
+  let show = function
+    | AssumeSuccess ->
+        "ana.pthread.assume_success"
+    | IgnoreAssigns ->
+        "ana.pthread.ignore_assign"
+
+
+  let eval = GobConfig.get_bool % show
+end
+
+(** [Function] module represents the supported pthread functions for the analysis *)
+module Function = struct
+  type t =
+    | Exit
+    | ThreadCreate
+    | ThreadJoin
+    | ThreadExit
+    | MutexInit
+    | MutexLock
+    | MutexUnlock
+    | CondVarInit
+    | CondVarBroadcast
+    | CondVarSignal
+    | CondVarWait
+
+  let funs =
+    [ ("exit", Exit)
+    ; ("pthread_create", ThreadCreate)
+    ; ("pthread_join", ThreadJoin)
+    ; ("pthread_exit", ThreadExit)
+    ; ("pthread_mutex_init", MutexInit)
+    ; ("pthread_mutex_lock", MutexLock)
+    ; ("pthread_mutex_unlock", MutexUnlock)
+    ; ("pthread_cond_init", CondVarInit)
+    ; ("pthread_cond_broadcast", CondVarBroadcast)
+    ; ("pthread_cond_signal", CondVarSignal)
+    ; ("pthread_cond_wait", CondVarWait)
+    ]
+
+
+  let error_code = -1
+
+  let supported = List.map fst funs
+
+  let from_string s = Option.map snd @@ List.find (( = ) s % fst) funs
+end
```

Comment:
This is somewhat duplicating `LibraryFunctions`.

Reply:
This is just a wrapper for easier pattern matching in the `special` analysis method. 
Do you propose extending `LibraryFunctions`? Since `LibraryFunctions` sees condition variable calls as lock/unlock, however, we wanna treat it differently here due to Promela Action extraction

Reply:
The classification part of `LibraryFunctions` doesn't classify `pthread_cond_*` as `Lock`/`Unlock` right now. It classifies them just as `Unknown`, so it seems sensible to just extend that classification in general:
https://github.com/goblint/analyzer/blob/cf11c4b0b3602f417a5573e56853b81940384998/src/analyses/libraryFunctions.ml#L8-L81

The only mentions of `pthread_cond_*` in `LibraryFunctions` just state that calling them doesn't invalidate any of their arguments, but that has nothing to do with locking/unlocking:
https://github.com/goblint/analyzer/blob/cf11c4b0b3602f417a5573e56853b81940384998/src/analyses/libraryFunctions.ml#L245-L248